### PR TITLE
More panel options

### PR DESCRIPTION
**Pull request by robin-carlier**
I finally found some time to implement what we discussed (and a bit more) in #8007

The main additions/modifications are

- A new `--layer` option for the kitten panel to specify on which layer the panel should be drawn.
- A new `--edge=none` option for the kitten, when set, the width, height and margins of the kitten can be freely set to get some flexibility in the geometry. When set, the panel leaves its exclusive_zone to `-1`, so it might get drawn on top of other things. Maybe we could also add a flag to conthol this as well?
- `--lines` and `--colums` are now split as two distinct parameters to control the size (in cells) when using `--edge=none`.
- New `--margin-...` options to set some margin if desired.
- New `focus-policy` flag.

With this, I think the kitten should be reasonably flexible to be used in fun ways.

I could test that the options behave as expected using Hyprland, but I don’t know if they do for other compositors.

### CODE REVIEW

File: kittens/panel/main.py [hunk 1]
```diff
@@ -39,6 +41,14 @@
 kitten.
 
 
+--layer
+choices=background, bottom, top, overlay
+default=bottom
+On a compositor that supports the wlr layer shell protocol, specifies the layer
```
Comment (kovidgoyal) [inline kittens/panel/main.py:82]: On a *Wayland* compositor to make it clear this option is Wayland specific

File: glfw/wl_window.c [hunk 2]
```diff
@@ -998,6 +998,10 @@ layer_set_properties(_GLFWwindow *window) {
                     panel_width = window->wl.width;
                     exclusive_zone = window->wl.width;
                     break;
+                case GLFW_EDGE_NONE:
+                    panel_width = window->wl.width;
+                    panel_width = window->wl.height;
```
Comment (kovidgoyal) [inline glfw/wl_window.c:1004]: Shouldnt this be panel height?

File: kittens/panel/main.py [hunk 3]
```diff
@@ -24,14 +25,20 @@
...
+
+--columns
+type=int
+default=1
+The number of columns shown in the panel if vertical. Ignored for background panels.
 
 
 --edge
-choices=top,bottom,left,right,background
+choices=top,bottom,left,right,background,none
```
Comment (kovidgoyal) [inline kittens/panel/main.py:68]: Add an explanation for the meaning of none to the help text

File: kittens/panel/main.py [hunk 4]
```diff
@@ -28,13 +28,37 @@
...
 type=int
 default=1
-The number of columns shown in the panel if vertical. Ignored for background panels.
+The number of columns shown in the panel. Ignored for background and horizontal panels.
+
+
+--margin-top
+type=int
+default=0
+Request a given top margin to the compositor.
```
Comment (kovidgoyal) [inline kittens/panel/main.py:46]: Add that the margin only works on Wayland compositor support layer shell. Add to help text for all margin options.

**Comment by kovidgoyal** [comment]
I am fine with adding a flag to control exclusive_zone.

File: glfw/wl_window.c (hunk 2 above)
Comment (robin-carlier) [reply glfw/wl_window.c]: Yes, good catch!

**Comment by robin-carlier** [comment]
> I am fine with adding a flag to control exclusive_zone.

I split it into two flags: `override-exclusive-zone` to allow override for panels anchored through e.g, `edge=top`, and a `exclusive-zone` flag to actually pass the value in case it is overriden.

**Comment by kovidgoyal** [comment]
Merged. It might be good to add some "recipes" or cookbook style
examples for doing common things to the docs for the panel kitten in
docs/kittens/panel.rst

**Pull request merged**